File: golang/src/detectors/log-injection/LogInjection.go
package main

import (
	"log"
	"net/http"
    "strings"
)

// {fact rule=log-injection@v1.0 defects=1}
func logInjectionNoncompliant(req *http.Request) {
	username := req.URL.Query()["username"][0]
    // Noncompliant: User provided input is passed directly to log.
	log.Printf("user %s logged in.\n", username)
}
// {/fact}

// {fact rule=log-injection@v1.0 defects=0}
func logInjectionCompliant(req *http.Request) {
	username := req.URL.Query()["username"][0]
	escapedUsername := strings.ReplaceAll(username, "\n", "")
	escapedUsername = strings.ReplaceAll(escapedUsername, "\r", "")
    // Compliant: User provided input is sanitized before passing to log.
	log.Printf("user %s logged in.\n", escapedUsername)
}
// {/fact}

File: ruby/src/detectors/log-injection/log-injection.rb
=begin
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
=end

require 'logger'

class UsersController < ApplicationController
  include ERB::Util

  def init_logger
    if @logger == nil
      @logger = Logger.new STDOUT
    end
  end
  
# {fact rule=log-injection@v1.0 defects=1}
  def log_params_noncompliant
    init_logger

    unsanitized = params[:foo]
    # Noncompliant: Unsanitized user-input is used in logger
    @logger.error "input: " + unsanitized 
  end
# {/fact}
  

# {fact rule=log-injection@v1.0 defects=0}
  def log_params_compliant
    init_logger

    unsanitized = params[:foo]

    sanitized = unsanitized.gsub("\n", "")
    # Compliant: Sanitized user-input is used in logger
    @logger.warn "input: " + sanitized 
  end
# {/fact}

end